### PR TITLE
feat: add bindings to Js.Set

### PR DESCRIPTION
**Pull request by tatchi**
@anmonteiro I couldn't push directly to your branch, so I created a new one with the bindings improvements

### CODE REVIEW

File: jscomp/test/js_set_test.ml
Comment (tatchi) [inline jscomp/test/js_set_test.ml:1]: Are these tests useful, or is it better to move them to `test/blackbox-tests/js_set.t` ? Seems like I couldn't make them to fail 🤔

File: jscomp/runtime/js_set.ml [hunk 1]
```diff
@@ -1,3 +1,34 @@
 (** ES6 Set API *)
 
 type 'a t
+
+external make : unit -> 'a t = "Set" [@@mel.new]
+external fromArray : 'a array -> 'a t = "Set" [@@mel.new]
+external toArray : 'a t -> 'a array = "Array.from"
+external size : 'a t -> int = "size" [@@mel.get]
+external add : value:'a -> 'a t = "add" [@@mel.send.pipe: 'a t]
```
Comment (tatchi) [inline jscomp/runtime/js_set.ml:9]: I'm not sure returning `'a t` is a good idea. I guess you did it to have a nice pipe API, but on the JS side the set is actually mutated, and this API doesn't make it obvious. See https://github.com/melange-re/melange/pull/1061/files#diff-64db361f42339ab2d557daeaff34549e0d79fc297dfc2411317c6d64f2e6557bR33-R41

File: test/blackbox-tests/js_set.t [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  > 
+  > let t_1 = Js.Set.make ()
+  > let t_2 = Js.Set.add ~value:1 t_1
+  > let () = Js.log (Printf.sprintf "t_1 size = %i; t_2 size = %i" (Js.Set.size t_1) (Js.Set.size t_2))
+  > EOF
+  $ dune build @melange
+  $ node _build/default/out/x.js
+  true
+  false
+  t_1 size = 1; t_2 size = 1
```
Comment (tatchi) [inline test/blackbox-tests/js_set.t:41]: I would not expect the size of `t_1` to be `1` with such API. Or am I missing something ?

File: jscomp/test/js_set_test.ml
Comment (tatchi) [reply jscomp/test/js_set_test.ml:1]: Looks like they're run on CI :)

File: jscomp/runtime/js_set.ml (hunk 1 above)
Comment (anmonteiro) [reply jscomp/runtime/js_set.ml:9]: Indeed, but the documentation mentions the return value is the set indeed. So we'd be lying by pretending it's `unit` even if that makes it more obvious that the set was mutated, no? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/add#return_value

File: jscomp/test/js_set_test.ml
Comment (anmonteiro) [reply jscomp/test/js_set_test.ml:1]: Ah, they'll be compiled and promoted if you comment this stanza https://github.com/melange-re/melange/blob/main/jscomp/dune.

File: test/blackbox-tests/js_set.t (hunk 2 above)
Comment (anmonteiro) [reply test/blackbox-tests/js_set.t:41]: depends whether you think that what you have is a reference or a value. In OCaml, `Set` and `Map` are values, while here when binding to JS it's a reference (like e.g.`ref` or `Queue.t` etc.) which is mutated underneath. So the test looks correct to me.

**Comment by anmonteiro** [comment]
> CI fails due to `.difference` not being defined. Seems like some of the bindings are very new (still in stage-2) https://www.proposals.es/proposals/New%20Set%20methods or https://caniuse.com/mdn-javascript_builtins_set_difference

Yeah, I mentioned this in https://github.com/melange-re/melange/pull/1047#issue-2109311733. It's fine to just comment them out for now.

**Comment by anmonteiro** [comment]
merged in #1047

**Pull request closed without merging**